src/shims/foreign_items.rs

```diff
@@ -359,6 +363,27 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
     ) -> InterpResult<'tcx, EmulateByNameResult<'mir, 'tcx>> {
         let this = self.eval_context_mut();
 
+        let tcx = this.tcx.tcx;
+        // First deal with any external C functions in linked .so file
+        // (if any SO file is specified).
+        if let Some(_) = this.machine.external_c_so_lib.as_ref() {
+            if let Some(local_id) = def_id.as_local() {
+                if let Some(extern_c_fct_rep) = ExternalCFuncDeclRep::from_hir_node(
+                    &tcx.hir().get(tcx.hir().local_def_id_to_hir_id(local_id)),
+                    link_name,
+                ) {
+                    // An error here means that the function being called was not successfully called
+                    // via the specified SO file; we should continue and check if it corresponds to
+                    // a provided shim.
+                    if let Ok(_) =
+                        this.call_and_add_external_c_fct_to_context(extern_c_fct_rep, dest, args)
+                    {
+                        return Ok(EmulateByNameResult::ExecutedExternalCCall);
+                    }
```

Comment:
I think we should error out if this failed. The only "soft failure" mode I'd expect would be that the shared object does not export this function. You could make `call_and_add_external_c_fct_to_context` return a `InterpResult<bool>` that is only Ok(`true`) if a function has been found (all failures should be an `Err`). Then you can check the boolean here
```suggestion
                    if this.call_and_add_external_c_fct_to_context(extern_c_fct_rep, dest, args)?
                    {
                        return Ok(EmulateByNameResult::ExecutedExternalCCall);
                    }
```

Reply:
Hmm, ok I see your point; and you're right that the main error case is when the SO doesn't export the function. Right now the issue is that since we're trying to call the function through the SO before we try the shims, if we error out for a function not being found in the SO then the shims will never be tried. There's 2 potential solutions I see:
1. don't error out if the function is not exported by the SO and continue to the shims execution 
2. try the shims before we try the function via the SO

For now I'll implement 1 but I'm happy to switch :)

Reply:
I have not read all of the discussion yet, but I do think the shims we have should usually take precedence. Or is there a case where having the FFI take precedence is important?

Reply:
Probably when we have a shim returning a file handle and then that handle goes to a FFI function. That will go horribly wrong

Reply:
I thought about solution 1 when I wrote my comment. It replicates what you have right now with little changes. I just don't want to hide the argument type mismatch errors, which your current logic does by "catching" errors here

Reply:
Yeah we have that file handle issue both ways though -- if a function is missing from the SO and we use the shim instead, things will still go horribly wrong.

Reply:
One small issue with both of these solutions that I'm currently running into is that the failing test output is now different when a SO is provided and the function being called is not found (since the error will say something like "function not exported in SO"). With my old code this didn't come up because with any error in using the SO would go forward and try with the shims, and if it still failed there the error message would be the same as it was originally.

Reply:
If the idea is that the SO *replaces* all shims, then that seems fine?

Don't run all the existing tests with the SO, that makes little sense then. Instead have a few tests dedicated at calling functions in the SO. The existing tests are for core Miri features and the Miri shims, after all, I wouldn't expect the SO to reproduce them -- Miri shims will always be better at detecting UB, for example.

Reply:
Ok that makes sense -- sounds like a plan.

Reply:
> One small issue with both of these solutions that I'm currently running into is that the failing test output is now different when a SO is provided and the function being called is not found (since the error will say something like "function not exported in SO").

This is why I would like for that not to be an error, but instead return `Ok(false)` or sth. Then you can choose to ignore it and continue to the shims.

Reply:
Good point -- in the current version of the code I'm checking the arguments [before trying to load the function from the SO file](https://github.com/emarteca/miri/blob/9db25a1b8d9fb31aa908eadebaf6ed87c1647699/src/c_ffi_support.rs#L179). Swapping the order of these computations fixes it.

However, rerunning all the tests with the SO file it looks like some functions seem to be implicitly exported by the SO file, like `malloc`. For example, [this test](https://github.com/emarteca/miri/blob/9db25a1b8d9fb31aa908eadebaf6ed87c1647699/tests/pass/function_calls/disable_abi_check.rs#L16) fails with the message:
```
>error: unsupported operation: Unsupported scalar argument type to external C function: $HEX
>  --> $DIR/disable_abi_check.rs:LL:CC
>   |
>LL |         let _ = malloc(0);
>   |                 ^^^^^^^^^ Unsupported scalar argument type to external C function: $HEX
```
Which is because the function _is_ retrieved from the SO file, but it then expects a hex argument which is not one of the types that's supported right now (although, that reminds me I need to add support for `usize` and `isize`).

Reply:
Oh fun. Uh... I have no good solution right now beyond silently returning `OK(false)` for unsupported things instead of erroring out. That said, since this is all experimental I'd be fine with falling back to "errors are silently ignored" for now, until there is no more unsupported api surface and all errors are true API mismatches